File: rust-backend/src/animation/animated_stack.rs
use std::fmt::Debug;

use super::{AnimationBuffer, AnimationBufferWrite, DataStructure, NewInstance, AnimatedFromIterator};

pub struct AnimatedStack<T: std::fmt::Debug> {
    contents: Vec<T>,
    pub buffer: AnimationBuffer,
}

impl<T: Debug> Debug for AnimatedStack<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.contents.fmt(f)
    }
}

impl<T: Debug> NewInstance<T> for AnimatedStack<T> {
    fn new(buffer: &AnimationBuffer) -> Self {
        Self {
            contents: Vec::new(),
            buffer: buffer.with_ds(DataStructure::Stack),
        }
    }
}

impl<T: Debug> AnimatedFromIterator<T> for AnimatedStack<T> {
    fn from_iter<I: IntoIterator<Item = T>>(buffer: &AnimationBuffer, iter: I) -> Self {
      Self{
          contents: Vec::from_iter(iter),
          buffer: buffer.with_ds(DataStructure::Stack),
      } 
    }
} 
#[allow(dead_code)]
impl<T: Debug> AnimatedStack<T> {
    /* pub fn new(buffer: &AnimationBuffer) -> AnimatedStack<T> {
        Self {
            contents: Vec::new(),
            buffer: buffer.with_ds(DataStructure::Stack),
        }
    } */
    pub fn push(&mut self, element: T) {
        self.buffer.write("push", &element);
        self.contents.push(element);
    }
    pub fn push_back(&mut self, element: T) {
        self.push(element);
    }
    pub fn pop(&mut self) -> Option<T> {
        let val = self.contents.pop();
        self.buffer.write("pop", &val);
        val
    }
    pub fn pop_back(&mut self) -> Option<T> {
        self.pop()
    }
    pub fn peek(&self) -> Option<&T> {
        let val = self.contents.last();
        self.buffer.write("peek", &val);
        val
    }
    pub fn peek_back(&self) -> Option<&T> {
        self.peek()
    }
    pub fn top(&self) -> Option<&T> {
        self.peek()
    }
    pub fn len(&self) -> usize {
        self.buffer.write("len", &self.contents.len());
        self.contents.len()
    }
    pub fn size(&self) -> usize {
        self.len()
    }
    pub fn is_empty(&self) -> bool {
        self.buffer.write("is_empty", &self.contents.is_empty());
        self.contents.is_empty()
    }
    pub fn clear(&mut self) {
        self.buffer.write("clear", &self.contents.len());
        self.contents.clear();
    }
}

File: python-backend/animation/animated_stack.py
from .animation_buffer import AnimationBuffer
import pyfastrand
from typing import Any
from collections import deque

class AnimatedStack:

    def __init__(self, buffer: AnimationBuffer = None) -> None:
        self.name: str = "stack"
        self.id: int = pyfastrand.pcg32()
        self.contents: deque = deque()
        self.buffer: AnimationBuffer = AnimationBuffer() if not buffer else buffer

    def push_back(self, element: Any) -> None:
        self.buffer.write(data_structure = self.name, operation="push", id = self.id, data = element)
        self.contents.append(element)

    def pop_back(self) -> Any:
        element = self.contents[-1]
        self.buffer.write(data_structure = self.name, operation="pop", id=self.id, data = element)
        return self.contents.pop()

    def peek_back(self) -> Any:
        element = self.contents[-1]
        self.buffer.write(data_structure=self.name, operation = "peek", id = self.id, data = element)
        return element

    def top(self) -> Any:
        return self.peek_back()

    def len(self) -> int:
        length = len(self.contents)
        self.buffer.write(data_structure=self.name, operation="len", id = self.id, data = length)
        return length
    
    def size(self) -> int:
        return self.len()
    
    def is_empty(self) -> bool:
        self.buffer.write(data_structure=self.name, operation = "is_empty", id = self.id, data = not self.contents)
        return not self.contents
    
    def clear(self) -> None:
        self.buffer.write(data_structure=self.name, operation = "clear", id = self.id, data = len(self.contents))
        self.contents.clear()


